### PR TITLE
Invitation Codes

### PR DESCRIPTION
**Pull request by MatthewTurk247**
# Invitation Codes

## :recycle: Current situation & Problem
There is a new collection in Firestore called invitation code bucket called `invitationCodes`. During onboarding, we sign the user in anonymously so we can check if they have a valid code. If they do, we assign that code to their new de-anonymized account and remove it from `invitationCodes` so it cannot be used again. Some kind of verification step like this hedges against spam and unapproved use.

## :gear: Release Notes 
- Adds an authentication step to onboarding whereby the user confirms against valid Firebase invitation codes that they are enrolled in the study.

## :books: Documentation
See [Firebase documentation](https://firebase.google.com/docs/auth/web/anonymous-auth).

## :white_check_mark: Testing
TBD.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@MatthewTurk247 I think invitation codes should not be stored in the storage bucket but in the firebase database. The storage bucket should only be used for large files like the consent document. That's probably the best use case we have for that now.

In addition, we are still missing the functionality in the cloud function like in the Spezi Study App: https://github.com/StanfordSpezi/SpeziStudyApplication.

This PR includes a few more things but demonstrates how we included a basic functionality in the app: https://github.com/StanfordSpezi/SpeziStudyApplication/commit/4f5e88df259f91e6b2bd44eb1fb77fde919dc2e2.

The final cloud function and current logic is found in the study app repo. It has a multi-study approach which makes it a bit more complicated but follows the same approach to validate an invitation code:
- Cloud function: https://github.com/StanfordSpezi/SpeziStudyApplication/blob/main/functions/index.js
- Invitation Code View: [https://github.com/StanfordSpezi/SpeziStudyApplication/blob/main/StudyApplication/[…]ng/OnboardingMechanisms/InvitationCode/InvitationCodeView.swift](https://github.com/StanfordSpezi/SpeziStudyApplication/blob/main/StudyApplication/StudyOnboarding/OnboardingMechanisms/InvitationCode/InvitationCodeView.swift)

I don’t think that we have to initially update SpeziOnboarding at this point, we can implement all that in the app but it might be eventually nice to move it out to one of the Spezi Package if needed. 
The Spezi Study Application importantly does not include the mapping of the anonymous account to the other login (e.g. username or password). I think we should be able to do this without any changes to SpeziAccount. 

In addition, it does not include the protection rule on SignUp to only allow this if you have a proper invitation code. This is not needed in the Spezi Study App as we don’t have any Accounts in there but will be necessary for PICS.
You can find more about these cloud functions here: https://firebase.google.com/docs/auth/extend-with-blocking-functions?gen=2nd

I would suggest to use the same function and structure as in the Study App. We have a functions folder that contains everything. The Firebase configuration (https://github.com/StanfordSpezi/SpeziStudyApplication/blob/main/firebase.json) includes that so you can test all the cloud functions locally in the emulator without ever needing to deploy them to the cloud project.

File: PAWS/ECGRecordings/ECGRecording.swift
Comment (PSchmiedmayer) [inline PAWS/ECGRecordings/ECGRecording.swift:21]: Thank you!

File: PAWS/Onboarding/InvitationCodeView.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeView.swift]: Please be sure to update this to reflect the headers of the PAWS project.

File: PAWS/Onboarding/InvitationCodeView.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeView.swift:148]: It would be great if you can add a preview to this view and modify it right in Xcode.

File: functions/index.js
Comment (PSchmiedmayer) [inline functions/index.js]: For a complete sign up to work we should also check that the user document exists and also contains an invitation code. Might even make sense to check that is the same one that was retrieved from the query snapshot to check that the elements are completely consistent as done with the function above.

File: functions/index.js
Comment (PSchmiedmayer) [inline functions/index.js:1]: Looks nice; thank you for mentioning the study app as the inspiration and linking to it.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
Thank you for the improvements @MatthewTurk247! I had a few smaller comments; should be small stuff that we can to to get it across the finish line.

I committed a few smaller things that I noticed during the review.

File: functions/index.js
Comment (MatthewTurk247) [reply functions/index.js]: Sounds good. I have added the logic for these checks and will let you know if the overall functionality works as expected in the emulator.

Comment (PSchmiedmayer) [reply functions/index.js]: Sounds great; thank you!

Comment (PSchmiedmayer) [reply functions/index.js]: It would be great to write a UI test around this to verify the behavior in the application.

We might need to seed some data in the Firebase Setup to do that.
You can inspect how wo do that in the Spezi Study application as well: https://github.com/StanfordSpezi/SpeziStudyApplication?tab=readme-ov-file#firebase-emulator-based-development--testing

**Comment by PSchmiedmayer** [comment]
@MatthewTurk247 Looks like some of the REUSE checks are still failing, you will most likely need to add `<...>.license` files for the remaining files: https://github.com/StanfordBDHG/PediatricAppleWatchStudy/actions/runs/8422279017/job/23061206134?pr=54#step:4:7

File: functions/index.js
Comment (MatthewTurk247) [reply functions/index.js]: Sure, I will give it a try. As a heads-up, the sign-up view seems to glitch out when I successfully create an account, at which point I have to close the app, open it again, and then use a newly generated invitation code to bypass that screen to log in with the credentials I already have. It could just be a quirk of the iOS simulator or Firebase emulator, but I am not sure.

File: functions/index.js
Comment (MatthewTurk247) [inline functions/index.js]: What do we think about checking whether the used invitation code belongs to the user who triggered this request, @PSchmiedmayer? Like so:
```js
if (!invitationCodeDoc.exists || (invitationCodeDoc.data().used && invitationCodeDoc.data().usedBy !== userId)) { ... }
```
That is, we throw an error in the event that either the invitation code does not exist, or, separately, the invitation code is already used, and the person who redeemed it is not the current user. However, if the person who redeemed said invitation code is the current user, then no error is thrown.

File: PAWS/Onboarding/InvitationCodeError.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeError.swift]: You should be able to generate a localized string like this:
```suggestion
            String(localized: ...)
```

File: functions/index.js
Comment (PSchmiedmayer) [reply functions/index.js]: Sure, sounds like a fair logic. In the end this case should never happen in a production use case as the sign up method should only be called a single time but this could be useful for testing.

**Comment by PSchmiedmayer** [comment]
@MatthewTurk247 This release should unblock your issue thanks to @Supereg: https://github.com/StanfordSpezi/SpeziFirebase/releases/tag/1.1.0

**Comment by MatthewTurk247** [comment]
@PSchmiedmayer, the onboarding flow is almost there, but a failed account operation error still shows up, even though linking appears to work just fine in Firebase emulator now. And if I tap the sign-up button again, the error message says that the account already exists. I have a feeling that one or two lines of code might be missing somewhere, or something needs to be removed, but I cannot quite put a finger on it.

File: .reuse/dep5
Comment (PSchmiedmayer) [inline .reuse/dep5]: I would suggest to remove this here and rather go for .license files for the .json files and headers in the files where we can add headers to ensure the we have a consistent style for all code bases.

File: PAWS/Onboarding/InvitationCodeError.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeError.swift]: Please use `String(localized: ` for all the localized elements.

File: PAWS/Onboarding/InvitationCodeError.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeError.swift]: Please use `String(localized: ` for all the localized elements.

File: PAWS/Onboarding/InvitationCodeView.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeView.swift]: Please properly indent functions within parameters:
```suggestion
                    primaryAction: {
                        guard validation.validateSubviews() else {
                            return
                        }
                    
                        await verifyOnboardingCode()
                    },
```

File: PAWS/Onboarding/InvitationCodeView.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeView.swift]: I would suggest a consistent parameter formatting and indentation here:
```suggestion
                    secondaryText: "I Already Have an Account",
                    secondaryAction: {
                        try Auth.auth().signOut()
                        onboardingNavigationPath.nextStep()
                    }
                )
```

File: PAWS/Onboarding/InvitationCodeView.swift
Comment (PSchmiedmayer) [inline PAWS/Onboarding/InvitationCodeView.swift]: Please properly indent this code by 4 spaces only, not 8.

File: PAWS/PAWSStandard.swift
Comment (PSchmiedmayer) [inline PAWS/PAWSStandard.swift]: We should remove the comment here.

Nevertheless, on a higher level: Could actually always compute this and never need to store a flag if we store when the user has enrolled in the study within the user document that is generated when the user signs up (and where we currently store the invitation code). 

If we store the date of sign up in the user document we can always calculate the `isAdult` flag in an environment object somewhere in the view hierarchy, but all of the might need to happen in #52.

File: PAWSUITests/AccountCreationTests.swift
Comment (PSchmiedmayer) [inline PAWSUITests/AccountCreationTests.swift:1]: Nice job with the UI tests, very good and overs a lot of the elements in the app!

File: functions/index.js
Comment (PSchmiedmayer) [inline functions/index.js:1]: Looks good; thank you for the multiple iterations @MatthewTurk247!

**Review by PSchmiedmayer — APPROVED** [review]
@MatthewTurk247 Thank you for all the additional improvements! The overall PR looks good to me; I only had a few smaller comments that would be great to be addressed before merging the PR. 

Feel free to merge the PR once all of them are addressed 👍

**Pull request merged**
